index.d.ts

```diff
@@ -0,0 +1,156 @@
+declare module 'solc' {
+  export type Primitive =
+    'bool' |
+    'string' |
+    'address' |
+    'uint8' |
+    'uint16' |
+    'uint32' |
+    'uint64' |
+    'uint128' |
+    'uint256' |
+    'int8' |
+    'int16' |
+    'int32' |
+    'int64' |
+    'int128' |
+    'int256' |
+    'bytes' |
+    'bytes20' |
+    'bytes32' |
+    'bool[]' |
+    'string[]' |
+    'address[]' |
+    'uint8[]' |
+    'uint16[]' |
+    'uint32[]' |
+    'uint64[]' |
+    'uint128[]' |
+    'uint256[]' |
+    'int8[]' |
+    'int16[]' |
+    'int32[]' |
+    'int64[]' |
+    'int128[]' |
+    'int256[]' |
+    'bytes[]' |
+    'bytes20[]' |
+    'bytes32[]';
+
+  export interface AbiParameter {
+    name: string,
+    type: Primitive,
+  }
+  
+  export interface AbiEventParameter extends AbiParameter {
+    indexed: boolean,
+  }
+  
+  export interface AbiFunction {
+    name: string,
+    type: 'function' | 'constructor' | 'fallback',
+    stateMutability: 'pure' | 'view' | 'payable' | 'nonpayable',
+    constant: boolean,
+    payable: boolean,
+    inputs: Array<AbiParameter>,
+    outputs: Array<AbiParameter>,
+  }
+  
+  export interface AbiEvent {
+    name: string,
+    type: 'event',
+    inputs: Array<AbiEventParameter>,
+    anonymous: boolean,
+  }
+  
+  export type Abi = Array<AbiFunction | AbiEvent>;
+  
+  interface CompilerInputSourceFile {
+    keccak256?: string;
+    urls: string[];
+  }
+  interface CompilerInputSourceCode {
+    keccak256?: string;
+    content: string;
+  }
+  interface CompilerInput {
+    language: "Solidity" | "serpent" | "lll" | "assembly";
+    settings?: any,
+    sources: {
+      [globalName: string]: CompilerInputSourceFile|CompilerInputSourceCode,
+    };
+  }
+  interface CompilerOutputError {
+    sourceLocation?: {
+      file: string;
+      start: number;
+      end: number;
+    };
+    type: "TypeError" | "InternalCompilerError" | "Exception";
+    component: "general" | "ewasm";
+    severity: "error" | "warning";
+    message: string;
+    formattedMessage?: string;
+  }
+  interface CompilerOutputEvmBytecode {
+    object?: string;
+    opcodes?: string;
+    sourceMap?: string;
+    linkReferences?: {} | {
+      [globalName: string]: {
+        [name: string]: {start: number, length: number}[];
+      };
+    };
+  }
+  interface CompilerOutputSources {
+    [globalName: string]: {
+      id: number;
+      ast?: any;
+      legacyAST?: any;
+    },
+  }
+  interface CompilerOutputContracts {
+    [globalName: string]: {
+      [contractName: string]: {
+        abi?: Abi;
+        metadata?: string;
+        userdoc?: any;
+        devdoc?: any;
+        ir?: string;
+        evm: {
```

Comment:
Even `evm` is optional, not only the fields below. Practically everything is optional in the response JSON.

Reply:
Fixed.

Reply:
Side note: I'm not a fan of responses where everything is optional, it makes the receiver have to litter their code with null coalescing which is painful in JavaScript.  However, that is out of scope of this PR which is just trying to match reality.